### PR TITLE
Support spring-kafka

### PR DESCRIPTION
**Pull request by zhaoyuguang**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues
#5244

### CODE REVIEW

**Comment by wu-sheng** [comment]
I agree with @kezhenxu94. 
@zhaoyuguang Could you explain the details about what is the background?

**Comment by zhaoyuguang** [comment]
At the kafka consumer end, it cannot be tracing with ` user code`. See this [FAQ:kafka-plugin](https://github.com/apache/skywalking/blob/master/docs/en/FAQ/kafka-plugin.md) for details. As for the hook method, it wants to be compatible with traces that did not receive or have no data before, because the consumer's logic is always while true

File: test/plugin/scenarios/spring-kafka-scenario/src/main/java/test/org/apache/skywalking/apm/testcase/spring/kafka/controller/CaseController.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        container.getContainerProperties().setAckMode(ContainerProperties.AckMode.MANUAL_IMMEDIATE);
+        container.start();
+    }
+
+    @RequestMapping("/spring-kafka-case")
+    @ResponseBody
+    public String springKafkaCase() {
+        kafkaTemplate.send(topicName, "key", "helloWorld");
+        return SUCCESS;
+    }
```
Comment (dmsolr) [inline test/plugin/scenarios/spring-kafka-scenario/src/main/java/test/org/apache/skywalking/apm/testcase/spring/kafka/controller/CaseController.java:108]: Is this an asynchronous operation?
```java
kafkaTemplate.send(topicName, "key", "helloWorld").get();
```

Comment (zhaoyuguang) [reply test/plugin/scenarios/spring-kafka-scenario/src/main/java/test/org/apache/skywalking/apm/testcase/spring/kafka/controller/CaseController.java:108]: no  is sync

**Comment by wu-sheng** [comment]
Your test case fails.

File: test/plugin/scenarios/spring-kafka-2.2.x-scenario/support-version.list [hunk 2]
```diff
@@ -14,16 +14,10 @@
...
+2.2.14.RELEASE
 2.2.13.RELEASE
+2.2.12.RELEASE
 2.2.11.RELEASE
+2.2.10.RELEASE
 2.2.9.RELEASE
-2.2.7.RELEASE
-2.2.5.RELEASE
-2.2.3.RELEASE
-2.1.14.RELEASE
```
Comment (wu-sheng) [inline test/plugin/scenarios/spring-kafka-2.2.x-scenario/support-version.list]: Aren't supporting 2.1.x? I think you should create a new test case?

File: apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/CallbackAdapterInterceptor.java [hunk 3]
```diff
@@ -45,8 +45,8 @@ public CallbackAdapterInterceptor(CallbackCache callbackCache) {
     public void onCompletion(RecordMetadata metadata, Exception exception) {
         ContextSnapshot snapshot = callbackCache.getSnapshot();
         AbstractSpan activeSpan = ContextManager.createLocalSpan("Kafka/Producer/Callback");
-        SpanLayer.asMQ(activeSpan);
         activeSpan.setComponent(ComponentsDefine.KAFKA_PRODUCER);
+        activeSpan.setLayer(SpanLayer.MQ);
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/CallbackAdapterInterceptor.java:49]: Why change this code style?

File: apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/CallbackInterceptor.java [hunk 4]
```diff
@@ -44,8 +44,8 @@ public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allAr
             ContextSnapshot snapshot = getSnapshot(cache);
             RecordMetadata metadata = (RecordMetadata) allArguments[0];
             AbstractSpan activeSpan = ContextManager.createLocalSpan("Kafka/Producer/Callback");
-            SpanLayer.asMQ(activeSpan);
             activeSpan.setComponent(ComponentsDefine.KAFKA_PRODUCER);
+            activeSpan.setLayer(SpanLayer.MQ);
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/CallbackInterceptor.java:48]: Same here.

File: apm-sniffer/apm-sdk-plugin/kafka-commons/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/SpringKafkaContext.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+package org.apache.skywalking.apm.plugin.kafka.define;
+
+public class SpringKafkaContext {
+
+    public SpringKafkaContext() {
+        needStop = false;
+    }
+
+    private boolean needStop;
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-commons/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/SpringKafkaContext.java:28]: Use @setter/@getter to follow the recommended code style.

File: apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/CallbackAdapterInterceptor.java (hunk 3 above)
Comment (zhaoyuguang) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/CallbackAdapterInterceptor.java:49]: There was a PR addition submitted by someone else the other day, but I had already done it earlier, so in the merge, it was in this style

File: test/plugin/scenarios/spring-kafka-2.2.x-scenario/support-version.list (hunk 2 above)
Comment (zhaoyuguang) [reply test/plugin/scenarios/spring-kafka-2.2.x-scenario/support-version.list]: I have seen that version 2.1.x, does not have this method, and i change `Supported-list.md` to 2.2.x

File: apm-sniffer/apm-sdk-plugin/kafka-commons/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/SpringKafkaContext.java (hunk 5 above)
Comment (zhaoyuguang) [reply apm-sniffer/apm-sdk-plugin/kafka-commons/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/SpringKafkaContext.java:28]: fixed

File: apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/CallbackAdapterInterceptor.java (hunk 3 above)
Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/CallbackAdapterInterceptor.java:49]: My point is, they are nearly the same. Why change it again?

**Comment by zhaoyuguang** [comment]
> FAQ doc about the Kafka should be updated too.

Please recheck

File: docs/en/FAQ/kafka-plugin.md [hunk 6]
```diff
@@ -5,4 +5,5 @@ The trace doesn't continue in kafka consumer side.
 The kafka client is pulling message from server, the plugin also just traces the pull action. As that, you need to do the manual instrument before the pull action, and include the further data process.
 
 ### Resolve
-Use Application Toolkit libraries to do manual instrumentation. such as `@Trace` annotation or OpenTracing API.
+Use Application Toolkit libraries to do manual instrumentation. such as `@Trace` annotation or OpenTracing API, Or if you're using `spring-kafka` 2.2.x or above, you can track the Consumer side without any code changes.
```
Comment (wu-sheng) [inline docs/en/FAQ/kafka-plugin.md:8]:
```suggestion
Use Application Toolkit libraries to do manual instrumentation. such as `@Trace` annotation or OpenTracing API, Or if you're using `spring-kafka` 2.2.x or above, you can track the Consumer side without any code change.
```

File: apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/CallbackAdapterInterceptor.java (hunk 3 above)
Comment (zhaoyuguang) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/CallbackAdapterInterceptor.java]: Np.  fixed

**Pull request merged**